tests/framework/e2e/etcdctl.go

```diff
@@ -507,3 +508,79 @@ func (ctl *EtcdctlV3) UserChangePass(user, newPass string) error {
 	_, err = cmd.Expect("Password updated")
 	return err
 }
+
+func (ctl *EtcdctlV3) RoleAdd(name string) (*clientv3.AuthRoleAddResponse, error) {
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "add", name), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleAddResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+func (ctl *EtcdctlV3) RoleGrantPermission(name string, key, rangeEnd string, permType clientv3.PermissionType) (*clientv3.AuthRoleGrantPermissionResponse, error) {
+	permissionType := authpb.Permission_Type_name[int32(permType)]
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "grant-permission", name, permissionType, key, rangeEnd), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleGrantPermissionResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+func (ctl *EtcdctlV3) RoleGet(role string) (*clientv3.AuthRoleGetResponse, error) {
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "get", role), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleGetResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+func (ctl *EtcdctlV3) RoleList() (*clientv3.AuthRoleListResponse, error) {
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "list"), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleListResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+func (ctl *EtcdctlV3) RoleRevokePermission(role string, key, rangeEnd string) (*clientv3.AuthRoleRevokePermissionResponse, error) {
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "revoke-permission", role, key, rangeEnd), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleRevokePermissionResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+func (ctl *EtcdctlV3) RoleDelete(role string) (*clientv3.AuthRoleDeleteResponse, error) {
+	line, err := SpawnCmdWithJsonPrinter(ctl.cmdArgs("role", "delete", role), nil)
+	if err != nil {
+		return nil, err
+	}
+	var resp clientv3.AuthRoleDeleteResponse
+	err = json.Unmarshal([]byte(line), &resp)
+	return &resp, err
+}
+
+// SpawnCmdWithJsonPrinter spawns an etcdctl process with provided command line arguments and environment variables
+// with json printer output and capture header line as output.
+func SpawnCmdWithJsonPrinter(args []string, envVars map[string]string) (string, error) {
+	args = append(args, "-w", "json")
+	cmd, err := SpawnCmd(args, envVars)
+	if err != nil {
+		return "", err
+	}
+	return cmd.ExpectFunc(func(txt string) bool {
+		// header substring is missing because of embedding in LeaseLeasesResponse.
+		// revision substring is missing because ClusterServer.header() does not fill revision in the header
+		return strings.Contains(txt, "header") || strings.Contains(txt, "revision")
+	})
+}
```

Comment:
```suggestion
func (ctl *EtcdctlV3) RoleDelete(role string) (*clientv3.AuthRoleDeleteResponse, error) {
    var resp clientv3.AuthRoleDeleteResponse
    err := ctl.spawnJsonCmd(&resp, "role", "delete",  role)
    return &resp, err
}

func (ctl *EtcdctlV3) spawnJsonCmd(output interface{}, args ...string) error {
	args = append(args, "-w", "json")
	cmd, err := SpawnCmd(append(ctl.cmdArgs(), args...), nil)
	if err != nil {
		return err
	}
	line, err := cmd.Expect("header")
	if err != nil {
		return err
	}
	err = json.Unmarshal([]byte(line), output)
	if err != nil {
		return err
	}
	return nil
}
```

`SpawnCmdWithJsonPrinter` you proposed doesn't work for any cmd, it depends directly on `-w json` parameters available only for some of the etcdctl commands. As so it should not be a function, but a method on EtcdctlV3 struct. 

I also proposed to make one additional step to include json parsing. It make sense that if we provide commands used to force json output, then we should always be able to parse json.